Generate Stream snapshots using next keyframe

uvjustin opened the pull request:
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Snapshots using stream will now use the next keyframe instead of the previous keyframe.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
`Stream` generates still images using keyframes from the video source. The default behavior for grabbing a still image is to use the previous keyframe, as this allows the function to return quickly e.g., when getting an image to display in the frontend.
However, it seems that using the previous keyframe does not work well when using the `Camera` snapshot service. Instead, users would prefer images generated from the next keyframe. This PR adds an option for `Stream` to generate images from the next keyframe, and it uses this when the `Camera` snapshot service is called.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #75930
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

uvjustin committed:
Add wait_for_next_keyframe option to stream images
Add STREAM_SNAPSHOT to CameraEntityFeature
Use wait_for_next_keyframe option for snapshots using stream

uvjustin committed:
Update stream test comments

uvjustin committed:
Add generic camera snapshot test

uvjustin commented on homeassistant/components/camera/__init__.py:
```diff
@@ -927,7 +928,11 @@ async def async_handle_snapshot_service(
         )
 
-    image = await camera.async_camera_image()
+    async with async_timeout.timeout(CAMERA_IMAGE_TIMEOUT):
```
The existing code doesn't seem to use a timeout for snapshots

allenporter replied:
Optional: You could also consider calling `_async_get_image` here to share timeout logic, but haven't thought this through.

uvjustin replied:
I'll take a look at that later, but I think it's probably better to avoid changing the PR more now.

uvjustin commented on homeassistant/components/camera/__init__.py:
```diff
@@ -102,6 +102,7 @@ class CameraEntityFeature(IntFlag):
 
     ON_OFF = 1
     STREAM = 2
+    STREAM_SNAPSHOT = 4
```
The `STREAM_SNAPSHOT `feature is just used to signal that the Camera subclass has an `async_camera_image` that takes a `wait_for_next_keyframe` argument

allenporter reviewed (commented):
Supporting camera snapshots from the next keyframe overall makes sense to me. Is the reason the camera snapshot needs this because it doesn't want a stale event for an automation? e.g. motion triggers, but the snapshot is before the motion



I would like to think more about about changes to the entity model here. Does this need an arch issue since its changing the entity model adding extra parameters?


One thought I have is that since you've codified the behavior with a feature, perhaps if the camera `CameraEntityFeature.STREAM_SNAPSHOT` the camera integration can do more of the work here not involving the other integrations. That is, camera can over the behavior of generating the snapshot from the stream, then only call into the integration for other types of snapshots. I wonder if that actually can decrease complexity around the integrations grabbing the stream object directly...

uvjustin commented:
> Supporting camera snapshots from the next keyframe overall makes sense to me. Is the reason the camera snapshot needs this because it doesn't want a stale event for an automation? e.g. motion triggers, but the snapshot is before the motion

Yes I believe so, although the difference between the previous and the next keyframe should only be a few seconds, those few seconds may result in a different scene. Also, I think some users use snapshots without using preload stream, so when the snapshot is fired it may not return anything if the stream was never active, or it can return a very stale image (from the last time the stream was active).

> I would like to think more about about changes to the entity model here. Does this need an arch issue since its changing the entity model adding extra parameters?

I was hoping to avoid any big arch changes, but yes, unfortunately the current structure means that this relatively minor change in `Stream` requires changes to the camera integrations and the camera platform itself.

> One thought I have is that since you've codified the behavior with a feature, perhaps if the camera `CameraEntityFeature.STREAM_SNAPSHOT` the camera integration can do more of the work here not involving the other integrations. That is, camera can over the behavior of generating the snapshot from the stream, then only call into the integration for other types of snapshots. I wonder if that actually can decrease complexity around the integrations grabbing the stream object directly...

Maybe we can play with that. I'm actually not sure if the feature is the best way to do it - open to other suggestions.

uvjustin commented:
> One thought I have is that since you've codified the behavior with a feature, perhaps if the camera `CameraEntityFeature.STREAM_SNAPSHOT` the camera integration can do more of the work here not involving the other integrations. That is, camera can over the behavior of generating the snapshot from the stream, then only call into the integration for other types of snapshots. I wonder if that actually can decrease complexity around the integrations grabbing the stream object directly...

Do you mean adding an extra service/adding a parameter to the snapshot service which allows the user to specify getting a stream snapshot? And then `camera` can go directly to stream instead of going through `async_camera_image`? That would make this PR no longer a breaking change, and it does make it a little cleaner in that we are not overriding the `async_camera_image` with a subclass method with additional parameters. One potential issue here is that may be some logic in the subclass `async_camera_image` that we would be skipping. For example, `generic` only uses a stream snapshot if there is no `still_image_url`, and `ONVIF` only uses a stream snapshot if `preload_stream` is active. However, by essentially exposing this "stream snapshot" as new service/parameter to the user, we may be able to assume they know what they are doing.

allenporter commented:
> > One thought I have is that since you've codified the behavior with a feature, perhaps if the camera `CameraEntityFeature.STREAM_SNAPSHOT` the camera integration can do more of the work here not involving the other integrations. That is, camera can over the behavior of generating the snapshot from the stream, then only call into the integration for other types of snapshots. I wonder if that actually can decrease complexity around the integrations grabbing the stream object directly...
> 
> Do you mean adding an extra service/adding a parameter to the snapshot service which allows the user to specify getting a stream snapshot? And then `camera` can go directly to stream instead of going through `async_camera_image`? That would make this PR no longer a breaking change, and it does make it a little cleaner in that we are not overriding the `async_camera_image` with a subclass method with additional parameters. One potential issue here is that may be some logic in the subclass `async_camera_image` that we would be skipping. For example, `generic` only uses a stream snapshot if there is no `still_image_url`, and `ONVIF` only uses a stream snapshot if `preload_stream` is active. However, by essentially exposing this "stream snapshot" as new service/parameter to the user, we may be able to assume they know what they are doing.


My thinking was:
- you have the new feature to show that snapshot streams are supported
- add a new method (somewhere) that handles the "snapshot from stream" behavior -- it can support a parameter for previous frame or next frame
- update the camera snapshot service `async_handle_snapshot_service` method to know about the "snapshot from stream" method if enabled with the "next frame" logic
- update `_async_get_image` to to know about the "snapshot from stream" method if enabled with the "previous frame ok" logic
- take all the logic out of all other integrations for using stream for a snapshot of all integrations
- for integrations like generic that already have a way to do still images, they don't enable snapshot from stream option.

My main thinking is that the camera integration already has everything it needs to support this without much help from the integrations really, since it gets everything right from the stream.

uvjustin commented:
Ok, that seems pretty much like what I thought you were saying. I'll make these changes.

uvjustin committed:
Get stream still images directly in camera
Remove getting stream images from generic, nest, and ONVIF
Refactor camera preferences
Add use_stream_for_stills setting to camera
Update tests

uvjustin committed:
Only attempt to get stream image if integration supports stream

uvjustin commented:
@allenporter I ended up removing the STREAM_SNAPSHOT feature flag and instead basing the decision whether or not to use stream for a snapshot on a new "use_stream_for_stills" camera setting stored in the entity registry.
This PR now has 3 main parts:
1) Removing the stream snapshot code from Generic Camera, Nest, and ONVIF
2) Adding a new camera setting and refactoring the camera preferences
3) Adding the new stream snapshot code to the camera platform
If this approach generally looks ok to you, I can break down the PR into those parts.

allenporter commented:
Yeah, that makes sense as a sequence of changes. I think `wait_for_next_keyframe` can be removed from the entity signature. Most integrations don't have settings objects but instead just have properties that are overridden (e.g. `_attr_foo` or a `foo` property) so i might suggest just having a new separate property for this.

uvjustin commented:
> Yeah, that makes sense as a sequence of changes. I think `wait_for_next_keyframe` can be removed from the entity signature. Most integrations don't have settings objects but instead just have properties that are overridden (e.g. `_attr_foo` or a `foo` property) so i might suggest just having a new separate property for this.

I agree that making this a setting adds some complexity, but the reason I made this a setting and not a property is because I did not think the decision whether or not to use this method could be automatically made by the integration (although it does seem like this is straightforward for Nest). For example, in Generic camera you can get the snapshot from the still image url or via stream. ONVIF is similar, but adds a third pathway for a still image using haffmpeg. Users might prefer to use the regular snapshot over the stream snapshot, or vice versa. I thought delegating this choice to the user would be the most appropriate.
However, on second thought I think your suggestion makes sense - we can just mimic the current behavior (automatically choosing a method) by having a property that makes the choice.
I'll make those changes.

uvjustin committed:
Use property instead of entity registry setting

uvjustin committed:
Split out getting stream prerequisites from stream_source in nest

uvjustin committed:
Use cached_property for rtsp live stream trait

uvjustin commented:
@allenporter Made those changes. Let me know what you think and if you think this PR still needs to be split up.

uvjustin committed:
Make rtsp live stream trait NestCamera attribute

allenporter commented on homeassistant/components/nest/camera.py:
```diff
@@ -205,9 +222,6 @@ async def async_camera_image(
         """Return bytes of camera image."""
         # Use the thumbnail from RTSP stream, or a placeholder if stream is
         # not supported (e.g. WebRTC)
```
```suggestion
        # not supported (e.g. WebRTC) as a fallback when 'use_stream_for_stills' if False
```

allenporter reviewed (commented):
Generally this looks great to me as the simplest possible API change to camera and it cleans up a bunch of interactions with the raw stream object which is great. I think it also needs updates to the developer documentation to add the attribute.  Adding this single attribute may not rise to the level of needing a full architecture discussion (i'm not sure), but generally think it would be good to have another member review too since its extending the entity model.

frenck commented:
There is a merge conflict that needs to be resolved. @uvjustin Could you take a look?

..Frenck

uvjustin committed:
Update homeassistant/components/nest/camera.py

Co-authored-by: Allen Porter <allen.porter@gmail.com>

uvjustin commented:
> There is a merge conflict that needs to be resolved. @uvjustin Could you take a look?
> 
> ..Frenck

Sure just resolved it. Will make a PR to add the `use_stream_for_stills` property on Camera to the developer documentation as @allenporter suggested

frenck commented:
Cool, @uvjustin 👍 

But now there is a mypy & ruff failure 😉 

../Frenck

uvjustin committed:
Change usage of async_timeout

uvjustin committed:
Change import formatting in generic/test_camera

uvjustin commented:
> Cool, @uvjustin 👍
> 
> But now there is a mypy & ruff failure 😉
> 
> ../Frenck

Whoops, fixed that and added the dev docs PR here: https://github.com/home-assistant/developers.home-assistant/pull/1888

uvjustin commented on homeassistant/components/nest/camera.py:
```diff
@@ -73,6 +74,25 @@ def __init__(self, device: Device) -> None:
+        """Check for stream prerequisites and return a CameraLiveStreamTrait or None."""
+        if not (self.supported_features & CameraEntityFeature.STREAM):
+            return None
+        if CameraLiveStreamTrait.NAME not in self._device.traits:
+            return None
```
@allenporter These lines are not covered by the tests. From looking at the code, because of the way `self.supported_features` is defined, it looks like this test is actually equivalent to the test in lines 81-82 above. Can we just remove these two lines here?

uvjustin replied:
There seems to be some overlap between `self._attr_is_streaming`, the `supported_features` property, and the additional check here in `_get_rtsp_live_stream_trait()`. Can these be combined and set in `__init__`. Is there a reason why the supported_features property can't be set through `_attr_supported_features`?

allenporter replied:
Setting `_attr_supported_features` up front sounds fine to me. This code was added before those attrs existed. Theoretically the device traits can change via pubsub messages, but if it did that would probably just be a bug anyway, so moving to _attr sounds better.

uvjustin replied:
Ok, I pushed a commit which just sets all of those things together in the `__init__`. Let me know if you'd prefer I submit the change to _attr as a separate PR.

uvjustin committed:
Simplify Nest camera property initialization
